### PR TITLE
Optimize SMT queries

### PR DESCRIPTION
**Pull request by MrChico**
Tries to minimize the time spent consulting SMT solvers, mainly by:
- Only caring about consistency of path conditions when in the interactive view; in headless mode dead paths will not be satisfiable in the final assertions anyway. This gives us the best of both worlds of the previous set up. This can avoid doing some queries completely.
- Use (smt) push and pop instead of `(reset-assertions)`. This allows the solvers to retain as much information as possible.

Some preliminary measures. For our current test suite, we have:
### cvc4 equivalence
```
Time taken (cvc4):
-------------------------------------------

real    0m44.033s
user    0m38.716s
sys     0m9.595s
``` 
### z3 equivalence
```

Time taken (z3):
-------------------------------------------

real    0m29.092s
user    0m23.985s
sys     0m8.733s
```
### cvc4 SMT-checker tests
```
Time Taken (cvc4):
---------------------------------


real    5m25.993s
user    5m19.739s
sys     0m29.593s
```

### z3 equivalence checking
```
Time taken (z3):
-------------------------------------------

real    0m39.792s
user    0m33.695s
sys     0m10.044s
```
### cvc4 equivalence checking
```
Time taken (cvc4):
-------------------------------------------

real    1m5.176s
user    0m58.971s
sys     0m10.816s
```

### cvc4 SMT-checker tests
```
Time Taken (cvc4):
---------------------------------


real    13m59.425s
user    13m43.175s
sys     0m41.843s
```

### CODE REVIEW

File: src/dapp-tests/integration/tests.sh [hunk 1]
```diff
@@ -59,12 +59,12 @@ test_hevm_symbolic() {
 
     solc --bin-runtime -o . --overwrite token.sol
     # This one explores all paths (cvc4 is better at this)
-    hevm symbolic --code $(<Token.bin-runtime) --solver cvc4
+    hevm symbolic --code $(<Token.bin-runtime) # --solver cvc4
```
Comment (xwvvvvwx) [inline src/dapp-tests/integration/tests.sh:62]: why commented?

**Review by xwvvvvwx — APPROVED** [review]
*very* nice :clap: 

Can we try enabling some of the more intense tests now things are faster?

File: src/dapp-tests/integration/tests.sh (hunk 1 above)
Comment (MrChico) [reply src/dapp-tests/integration/tests.sh:62]: mistake

Comment (MrChico) [inline src/dapp-tests/integration/tests.sh:62]:
```suggestion
    hevm symbolic --code $(<Token.bin-runtime) --solver cvc4
```

File: src/dapp-tests/integration/tests.sh [hunk 2]
```diff
@@ -59,12 +59,12 @@ test_hevm_symbolic() {
...
     rm -rf Token.bin-runtime
 
     # The contracts A and B should be equivalent:
     solc --bin-runtime -o . --overwrite AB.sol
-    hevm equivalence --code-a $(<A.bin-runtime) --code-b $(<B.bin-runtime) --solver cvc4
+    hevm equivalence --code-a $(<A.bin-runtime) --code-b $(<B.bin-runtime) #--solver cvc4
```
Comment (MrChico) [inline src/dapp-tests/integration/tests.sh:67]:
```suggestion
    hevm equivalence --code-a $(<A.bin-runtime) --code-b $(<B.bin-runtime) --solver cvc4
```

**Pull request merged**